commands/binding-tool/main.go

```diff
@@ -0,0 +1,244 @@
+	}
+
+	b, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+
+	wsClient.SendWhisperMessages(podDID, 0, [][]byte{b})
+	r := <-respCh
+
+	var bitcoindResp json.RawMessage
+	if err := json.Unmarshal(r.Content, &bitcoindResp); err != nil {
+		return nil, err
+	}
+
+	return bitcoindResp, nil
+}
+
+func main() {
+	var configFile string
+	flag.StringVar(&configFile, "c", "./config.yaml", "[optional] path of configuration file")
+	flag.StringVar(&configFile, "config", "./config.yaml", "[optional] path of configuration file")
+	flag.Usage = func() {
+		fmt.Fprintf(os.Stderr, "Usage: binding-tool [options] [command]\n")
+		flag.PrintDefaults()
+		fmt.Fprintf(os.Stderr, "Command:\n")
+		fmt.Fprintf(os.Stderr, "  bind \t\t\t\t\t initiate a bind process\n")
+		fmt.Fprintf(os.Stderr, "  bind_ack [nonce] \t\t\t respond a bind ack with a given nonce\n")
+		fmt.Fprintf(os.Stderr, "  bitcoind [JSONRPC request body] \t call bitcoind command\n")
+	}
+	flag.Parse()
+
+	commands := flag.Args()
+	if len(commands) == 0 {
+		flag.Usage()
+		return
+	}
+
+	config.LoadConfig(configFile)
+
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt)
+
+	messagingClient := messaging.New(http.DefaultClient,
+		viper.GetString("messaging.server_url"),
+		viper.GetString("messaging.client_jwt"),
+		viper.GetString("messaging.store.leveldb_path"),
+	)
+
+	if err := messagingClient.RegisterAccount(); err != nil {
+		log.WithError(err).Fatalf("registering account, error")
+	}
+
+	if err := messagingClient.RegisterKeys(); err != nil {
+		log.WithError(err).Fatalf("failed to register keys")
+	}
+
+	wsClient, err := messagingClient.NewWSClient()
+	if err != nil {
+		log.WithError(err).Panic("fail to start a websocket connection")
+	}
+	defer wsClient.Close()
+
+	done := make(chan struct{})
+
+	msgCh := wsClient.WhisperMessages()
+
+	privateKey, err := hex.DecodeString("61900f1b361bda95b318a219efda0685c9209f703f8638c269506860271c26d9")
+	if err != nil {
+		panic("unable to read private key")
+	}
+
+	go func() {
+		podDID := viper.GetString("pod.identity")
+
+		switch commands[0] {
+		case "bind":
+			nonce, err := bind(wsClient, msgCh, podDID)
+			if err != nil {
+				log.WithError(err).Error("bind request fail")
+				os.Exit(1)
+			}
+			log.WithField("podDID", podDID).WithField("nonce", nonce).Info("bind ok")
+		case "bind_ack":
+			if len(commands) != 2 {
+				flag.Usage()
+				break
+			}
+			if err := bindACK(wsClient, msgCh, podDID, commands[1], privateKey); err != nil {
+				log.WithError(err).Error("bind ack fail")
+				os.Exit(1)
+			}
+			log.Info("bind ack ok")
+		case "bitcoind":
+			if len(commands) != 2 {
+				flag.Usage()
+				break
+			}
+
+			resp, err := bitcoinCommand(wsClient, msgCh, podDID, commands[1])
+			if err != nil {
+				log.WithError(err).Panic("bitcoin request fail")
+				os.Exit(1)
+			}
+
+			log.WithField("resp", string(resp)).Info("bitcoin response")
+		default:
+			flag.Usage()
+		}
+
+		os.Exit(0)
```

Comment:
should this be wsclient.close(); close(done) ?

Reply:
I tried `close(done)` before. It prints out a lot connection logs. Since this is a tool simply to get response back from the server, I think it can terminate directly and let the os deal with the connection recycling.

Reply:
:smile_cat: